### ISSUE TITLE
Fluentd not connecting to Elasticsearch

### ISSUE THREAD

**Issue by jashykm** [fluent/fluentd-kubernetes-daemonset#523]
Hello,
I am trying to build an EFK stack and facing issues with Fluentd. Fluentd is not connecting to Elasticsearch and there are no errors in the fluentd pod logs.  Following are the details.
K8s version - 1.19
Container Runtime - cri-o
Elasticsearch installed using helm -
https://github.com/elastic/helm-charts/tree/master/elasticsearch

Used the this yaml file for fluentd with the image fluentd-kubernetes-daemonset:v1.11-debian-elasticsearch7-1 -  https://github.com/fluent/fluentd-kubernetes-daemonset/blob/master/fluentd-daemonset-elasticsearch-rbac.yaml 

netstat from the fluentd  pod shows TIME_WAIT for Elasticsearch for some time and the entry disappears 

root@fluentd-c9bqt:/home/fluent# netstat -ntp
Active Internet connections (w/o servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 172.16.122.254:39476    10.96.0.1:443           ESTABLISHED 6/ruby     
tcp        0      0 172.16.122.254:36854    172.16.122.255:9200     TIME_WAIT   -          
tcp        0      0 172.16.122.254:39472    10.96.0.1:443           ESTABLISHED 6/ruby     
tcp        0      0 172.16.122.254:39584    10.96.0.1:443           ESTABLISHED 23/ruby    
tcp        0      0 172.16.122.254:39582    10.96.0.1:443           ESTABLISHED 23/ruby

Following is the fluentd pod log:

```log
021-01-15 15:19:45 +0000 [info]: parsing config file is succeeded path="/fluentd/etc/fluent.conf"
2021-01-15 15:19:45 +0000 [info]: gem 'fluent-plugin-concat' version '2.4.0'
2021-01-15 15:19:45 +0000 [info]: gem 'fluent-plugin-dedot_filter' version '1.0.0'
2021-01-15 15:19:45 +0000 [info]: gem 'fluent-plugin-detect-exceptions' version '0.0.13'
2021-01-15 15:19:45 +0000 [info]: gem 'fluent-plugin-elasticsearch' version '4.3.3'
2021-01-15 15:19:45 +0000 [info]: gem 'fluent-plugin-grok-parser' version '2.6.2'
2021-01-15 15:19:45 +0000 [info]: gem 'fluent-plugin-json-in-json-2' version '1.0.2'
2021-01-15 15:19:45 +0000 [info]: gem 'fluent-plugin-kubernetes_metadata_filter' version '2.5.2'
2021-01-15 15:19:45 +0000 [info]: gem 'fluent-plugin-multi-format-parser' version '1.0.0'
2021-01-15 15:19:45 +0000 [info]: gem 'fluent-plugin-prometheus' version '1.6.1'
2021-01-15 15:19:45 +0000 [info]: gem 'fluent-plugin-record-modifier' version '2.0.1'
2021-01-15 15:19:45 +0000 [info]: gem 'fluent-plugin-rewrite-tag-filter' version '2.2.0'
2021-01-15 15:19:45 +0000 [info]: gem 'fluent-plugin-systemd' version '1.0.2'
2021-01-15 15:19:45 +0000 [info]: gem 'fluentd' version '1.11.5'
2021-01-15 15:19:47 +0000 [info]: using configuration file: <ROOT>
  
    @type systemd
    @id in_systemd_kubelet
    matches [{"_SYSTEMD_UNIT":"kubelet.service"}]
    read_from_head true
    tag "kubelet"
    <storage>
      @type "local"
      persistent true
      path "/var/log/fluentd-journald-kubelet-cursor.json"
    </storage>
    <entry>
      fields_strip_underscores true
    </entry>
  </source>
 ...............................
................................
    

  <filter kubernetes.**>
    @type kubernetes_metadata
    @id filter_kube_metadata
    kubernetes_url "https://10.96.0.1:443/api"
    verify_ssl true
    ca_file ""
    skip_labels false
    skip_container_metadata false
    skip_master_url false
    skip_namespace_metadata false
    watch true
  </filter>
  <match **>
    @type elasticsearch
    @id out_es
    @log_level "info"
    include_tag_key true
    host "elasticsearch-master-headless"
    port 9200
    path ""
    scheme http
    ssl_verify true
    ssl_version TLSv1_2
    user "elastic"
    password xxxxxx
    reload_connections false
    reconnect_on_error true
    reload_on_failure true
    log_es_400_reason false
    logstash_prefix "logstash"
    logstash_dateformat "%Y.%m.%d"
    logstash_format true
    index_name "logstash"
    target_index_key 
    type_name "fluentd"
    include_timestamp false
    template_name 
    template_file 
    template_overwrite false
    sniffer_class_name "Fluent::Plugin::ElasticsearchSimpleSniffer"
    request_timeout 5s
    suppress_type_name true
    enable_ilm false
    ilm_policy_id logstash-policy
    ilm_policy {}
    ilm_policy_overwrite false
    <buffer>
      flush_thread_count 8
      flush_interval 5s
      chunk_limit_size 2M
      queue_limit_length 32
      retry_max_interval 30
      retry_forever true
    </buffer>
  </match>
</ROOT>
2021-01-15 15:19:47 +0000 [info]: starting fluentd-1.11.5 pid=6 ruby="2.6.6"
2021-01-15 15:19:47 +0000 [info]: spawn command to main:  cmdline=["/usr/local/bin/ruby", "-Eascii-8bit:ascii-8bit", "/fluentd/vendor/bundle/ruby/2.6.0/bin/fluentd", "-c", "/fluentd/etc/fluent.conf", "-p", "/fluentd/plugins", "--gemfile", "/fluentd/Gemfile", "-r", "/fluentd/vendor/bundle/ruby/2.6.0/gems/fluent-plugin-elasticsearch-4.3.3/lib/fluent/plugin/elasticsearch_simple_sniffer.rb", "--under-supervisor"]
2021-01-15 15:19:47 +0000 [info]: adding match in @FLUENT_LOG pattern="fluent.**" type="null"
2021-01-15 15:19:47 +0000 [info]: adding filter pattern="kubernetes.**" type="kubernetes_metadata"
2021-01-15 15:19:47 +0000 [info]: adding match pattern="**" type="elasticsearch"
2021-01-15 15:19:47 +0000 [info]: adding source type="systemd"
2021-01-15 15:19:47 +0000 [info]: adding source type="systemd"
2021-01-15 15:19:47 +0000 [info]: adding source type="systemd"
2021-01-15 15:19:47 +0000 [info]: adding source type="prometheus"
2021-01-15 15:19:48 +0000 [info]: adding source type="prometheus_output_monitor"
2021-01-15 15:19:48 +0000 [info]: adding source type="tail"
2021-01-15 15:19:48 +0000 [info]: adding source type="tail"
2021-01-15 15:19:48 +0000 [info]: adding source type="tail"
2021-01-15 15:19:48 +0000 [info]: adding source type="tail"
2021-01-15 15:19:48 +0000 [info]: adding source type="tail"
2021-01-15 15:19:48 +0000 [info]: adding source type="tail"
2021-01-15 15:19:48 +0000 [info]: adding source type="tail"
2021-01-15 15:19:48 +0000 [info]: adding source type="tail"
2021-01-15 15:19:48 +0000 [info]: adding source type="tail"
2021-01-15 15:19:48 +0000 [info]: adding source type="tail"
2021-01-15 15:19:48 +0000 [info]: adding source type="tail"
2021-01-15 15:19:48 +0000 [info]: adding source type="tail"
2021-01-15 15:19:48 +0000 [info]: adding source type="tail"
2021-01-15 15:19:48 +0000 [info]: adding source type="tail"
2021-01-15 15:19:48 +0000 [info]: #0 starting fluentd worker pid=17 ppid=6 worker=0
2021-01-15 15:19:48 +0000 [info]: #0 [in_tail_container_logs] following tail of /var/log/containers/calico-node-mrmjb_kube-system_calico-node-aa1988509a815ea9ca36d5c894b460d44607d8e9acf8ac1daec025234ffe581c.log
2021-01-15 15:19:48 +0000 [warn]: #0 [in_tail_container_logs] invalid line found file="/var/log/containers/calico-node-mrmjb_kube-system_calico-node-aa1988509a815ea9ca36d5c894b460d44607d8e9acf8ac1daec025234ffe581c.log" line="2021-01-15T19:19:22.091622789+04:00 stdout F 2021-01-15 15:19:22.091 [INFO][40] felix/int_dataplane.go 1300: Applying dataplane updates" error="invalid time format: value = 2021-01-15T19:19:22.091622789+04:00, error_class = ArgumentError, error = string doesn't match"
2021-01-15 15:19:48 +0000 [warn]: #0 [in_tail_container_logs] invalid line found file="/var/log/containers/calico-node-mrmjb_kube-system_calico-node-aa1988509a815ea9ca36d5c894b460d44607d8e9acf8ac1daec025234ffe581c.log" line="2021-01-15T19:19:22.091622789+04:00 stdout F 2021-01-15 15:19:22.091 [INFO][40] felix/ipsets.go 223: Asked to resync with the dataplane on next update. family=\"inet\"" error="invalid time format: value = 2021-01-15T19:19:22.091622789+04:00, error_class = ArgumentError, error = string doesn't match"
```
.....................
........ and goes on
   

I can connect to the Elastic pod manually

```console
$ curl http://172.16.122.255:9200
{
  "name" : "elasticsearch-master-0",
  "cluster_name" : "elasticsearch",
  "cluster_uuid" : "OTVJdfgsTyyw2Rt0TEGVNA",
  "version" : {
    "number" : "7.10.2",
    "build_flavor" : "default",
    "build_type" : "docker",
    "build_hash" : "747e1cc71def077253878a59143c1f785afa92b9",
    "build_date" : "2021-01-13T00:42:12.435326Z",
    "build_snapshot" : false,
    "lucene_version" : "8.7.0",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}
```

**Comment by jashykm** [comment]
Following are the entries from the log after I used a configmap for the fluentd.conf file and enabled debug.  Any help to fix the issue would be appreciated. 

2021-01-15 23:42:34 +0000 [info]: starting fluentd-1.11.5 pid=6 ruby="2.6.6"
2021-01-15 23:42:34 +0000 [info]: spawn command to main:  cmdline=["/usr/local/bin/ruby", "-Eascii-8bit:ascii-8bit", "/fluentd/vendor/bundle/ruby/2.6.0/bin/fluentd", "-c", "/fluentd/etc/fluent.conf", "-p", "/fluentd/plugins", "--gemfile", "/fluentd/Gemfile", "-r", "/fluentd/vendor/bundle/ruby/2.6.0/gems/fluent-plugin-elasticsearch-4.3.3/lib/fluent/plugin/elasticsearch_simple_sniffer.rb", "--under-supervisor"]
2021-01-15 23:42:35 +0000 [info]: adding match pattern="fluent.**" type="null"
2021-01-15 23:42:35 +0000 [info]: adding match pattern="**" type="elasticsearch"
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] 'host elasticsearch-master' is tested built-in placeholder(s) but there is no valid placeholder(s). error: Parameter 'host: elasticsearch-master' doesn't have tag placeholder
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] 'index_name logstash' is tested built-in placeholder(s) but there is no valid placeholder(s). error: Parameter 'index_name: logstash' doesn't have tag placeholder
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] 'template_name ' is tested built-in placeholder(s) but there is no valid placeholder(s). error: Parameter 'template_name: ' doesn't have tag placeholder
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] 'logstash_prefix logstash' is tested built-in placeholder(s) but there is no valid placeholder(s). error: Parameter 'logstash_prefix: logstash' doesn't have tag placeholder
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] 'logstash_dateformat %Y.%m.%d' is tested built-in placeholder(s) but there is no valid placeholder(s). error: Parameter 'logstash_dateformat: %Y.%m.%d' has timestamp placeholders, but chunk key 'time' is not configured
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] 'logstash_dateformat %Y.%m.%d' is tested built-in placeholder(s) but there is no valid placeholder(s). error: Parameter 'logstash_dateformat: %Y.%m.%d' doesn't have tag placeholder
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] 'deflector_alias ' is tested built-in placeholder(s) but there is no valid placeholder(s). error: Parameter 'deflector_alias: ' doesn't have tag placeholder
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] 'application_name default' is tested built-in placeholder(s) but there is no valid placeholder(s). error: Parameter 'application_name: default' doesn't have tag placeholder
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] 'ilm_policy_id logstash-policy' is tested built-in placeholder(s) but there is no valid placeholder(s). error: Parameter 'ilm_policy_id: logstash-policy' doesn't have tag placeholder
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] Need substitution: false
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] 'host_placeholder elasticsearch-master' is tested built-in placeholder(s) but there is no valid placeholder(s). error: Parameter 'host_placeholder: elasticsearch-master' doesn't have tag placeholder
**2021-01-15 23:42:35 +0000 [warn]: #0 **[out_es] Detected ES 7.x: `_doc` will be used as the document `_type`.****
2021-01-15 23:42:35 +0000 [info]: adding source type="tail"
2021-01-15 23:42:35 +0000 [warn]: #0 define <match fluent.**> to capture fluentd logs in top level is deprecated. Use <label @FLUENT_LOG> instead
2021-01-15 23:42:35 +0000 [info]: #0 starting fluentd worker pid=14 ppid=6 worker=0
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] buffer started instance=69931600986520 stage_size=0 queue_size=0
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] flush_thread actually running
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] flush_thread actually running
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] flush_thread actually running
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] flush_thread actually running
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] flush_thread actually running
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] flush_thread actually running
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] flush_thread actually running
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] flush_thread actually running
2021-01-15 23:42:35 +0000 [debug]: #0 [out_es] enqueue_thread actually running
2021-01-15 23:42:35 +0000 [info]: #0 [in_tail_container_logs] following tail of /var/log/containers/calico-node-mrmjb_kube-system_calico-node-aa1988509a815ea9ca36d5c894b460d44607d8e9acf8ac1daec025234ffe581c.log
2021-01-15 23:42:35 +0000 [warn]: #0 [in_tail_container_logs] invalid line found file="/var/log/containers/calico-node-mrmjb_kube-system_calico-node-aa1988509a815ea9ca36d5c894b460d44607d8e9acf8ac1daec025234ffe581c.log" line="2021-01-16T03:42:24.044722765+04:00 stdout F 2021-01-15 23:42:24.043 [INFO][38] felix/calc_graph.go 409: Local endpoint deleted id=WorkloadEndpoint(node=master.kube.local, orchestrator=k8s, workload=kube-logging/fluentd-ghzkt, name=eth0)" error="invalid time format: value = 2021-01-16T03:42:24.044722765+04:00, error_class = ArgumentError, error = string doesn't match"
2021-01-15 23:42:35 +0000 [warn]: #0 [in_tail_container_logs] invalid line found file="/var/log/containers/calico-node-mrmjb_kube-system_calico-node-aa1988509a815ea9ca36d5c894b460d44607d8e9acf8ac1daec025234ffe581c.log" line="2021-01-16T03:42:24.083352324+04:00 stdout F 2021-01-15 23:42:24.045 [INFO][38] felix/int_dataplane.go 1172: Received *proto.WorkloadEn:

**Comment by agup006** [comment]
Hmm I believe this format is CRI parser related @repeatedly do you have Fluentd configuration that could be used for this?

**Comment by jashykm** [comment]
**Here is the fluentd.conf** .....

@include kubernetes.conf
<match **>
   @type elasticsearch
   @id out_es
   @log_level debug
   include_tag_key true
   host "#{ENV['FLUENT_ELASTICSEARCH_HOST']}"
   port "#{ENV['FLUENT_ELASTICSEARCH_PORT']}"
   path "#{ENV['FLUENT_ELASTICSEARCH_PATH']}"
   scheme "#{ENV['FLUENT_ELASTICSEARCH_SCHEME'] || 'http'}"
   ssl_verify "#{ENV['FLUENT_ELASTICSEARCH_SSL_VERIFY'] || 'true'}"
   ssl_version "#{ENV['FLUENT_ELASTICSEARCH_SSL_VERSION'] || 'TLSv1_2'}"
   user "#{ENV['FLUENT_ELASTICSEARCH_USER'] || use_default}"
   password "#{ENV['FLUENT_ELASTICSEARCH_PASSWORD'] || use_default}"
   reload_connections "#{ENV['FLUENT_ELASTICSEARCH_RELOAD_CONNECTIONS'] || 'false'}"
   reconnect_on_error "#{ENV['FLUENT_ELASTICSEARCH_RECONNECT_ON_ERROR'] || 'true'}"
   reload_on_failure "#{ENV['FLUENT_ELASTICSEARCH_RELOAD_ON_FAILURE'] || 'true'}"
   log_es_400_reason "#{ENV['FLUENT_ELASTICSEARCH_LOG_ES_400_REASON'] || 'false'}"
   logstash_prefix "#{ENV['FLUENT_ELASTICSEARCH_LOGSTASH_PREFIX'] || 'logstash'}"
   logstash_dateformat "#{ENV['FLUENT_ELASTICSEARCH_LOGSTASH_DATEFORMAT'] || '%Y.%m.%d'}"
   logstash_format "#{ENV['FLUENT_ELASTICSEARCH_LOGSTASH_FORMAT'] || 'true'}"
   index_name "#{ENV['FLUENT_ELASTICSEARCH_LOGSTASH_INDEX_NAME'] || 'logstash'}"
   type_name "#{ENV['FLUENT_ELASTICSEARCH_LOGSTASH_TYPE_NAME'] || 'fluentd'}"
   include_timestamp "#{ENV['FLUENT_ELASTICSEARCH_INCLUDE_TIMESTAMP'] || 'false'}"
   template_name "#{ENV['FLUENT_ELASTICSEARCH_TEMPLATE_NAME'] || use_nil}"
   template_file "#{ENV['FLUENT_ELASTICSEARCH_TEMPLATE_FILE'] || use_nil}"
   template_overwrite "#{ENV['FLUENT_ELASTICSEARCH_TEMPLATE_OVERWRITE'] || 'true'}"
   sniffer_class_name "#{ENV['FLUENT_SNIFFER_CLASS_NAME'] || 'Fluent::Plugin::ElasticsearchSimpleSniffer'}"
   request_timeout "#{ENV['FLUENT_ELASTICSEARCH_REQUEST_TIMEOUT'] || '5s'}"
   <buffer>
     flush_thread_count "#{ENV['FLUENT_ELASTICSEARCH_BUFFER_FLUSH_THREAD_COUNT'] || '8'}"
     flush_interval "#{ENV['FLUENT_ELASTICSEARCH_BUFFER_FLUSH_INTERVAL'] || '5s'}"
     chunk_limit_size "#{ENV['FLUENT_ELASTICSEARCH_BUFFER_CHUNK_LIMIT_SIZE'] || '2M'}"
     queue_limit_length "#{ENV['FLUENT_ELASTICSEARCH_BUFFER_QUEUE_LIMIT_LENGTH'] || '32'}"
     retry_max_interval "#{ENV['FLUENT_ELASTICSEARCH_BUFFER_RETRY_MAX_INTERVAL'] || '30'}"
     retry_forever true
   </buffer>
</match>
...........................

**and Kubernetes.conf.**
<match fluent.**>
  @type null
</match>
<source>
  @type tail
  @id in_tail_container_logs
  path "/var/log/containers/*.log"
  exclude_path ["/var/log/containers/fluentd*"]
  pos_file "/var/log/fluentd-{log name}.log.pos"
  tag "kubernetes.*"
  read_from_head true
  format /^(?<time>.+) (?<stream>stdout|stderr)( (?<logtag>.))? (?<log>.*)$/
  time_format "%Y-%m-%dT%H:%M:%S.%NZ"
  </source>

**Comment by jashykm** [comment]
Hi, did you get a chance to take a look at the files?
Thanks,

**Comment by jashykm** [comment]
I just cannot make it work even with the new image v1.12.0-debian-elasticsearch7-1.0. I thought the CRI parser isssue has been fixed in the latest. I can't figure out whats wrong!

**Comment by cosmo0920** [comment]
> I just cannot make it work even with the new image v1.12.0-debian-elasticsearch7-1.0

Your using images does not include CRI parser feature.
CRI parser feature is included since  v1.12.0-debian-elasticsearch7-1.1.

Using CRI parser feature, you should specify the environment variable as follows:

* `FLUENT_CONTAINER_TAIL_PARSER_TYPE` as `cri`
* https://github.com/fluent/fluentd-kubernetes-daemonset/blob/0b7485e5abf8739e51e9b68d6a5c0f20e3f58270/docker-image/v1.12/debian-elasticsearch7/conf/tail_container_parse.conf
* https://github.com/fluent/fluentd-kubernetes-daemonset/blob/0b7485e5abf8739e51e9b68d6a5c0f20e3f58270/docker-image/v1.12/debian-elasticsearch7/conf/kubernetes.conf#L19
